### PR TITLE
Feature/Echo Installer

### PR DESCRIPTION
**Pull request by Strepto**
Note: This repository is open.

### CODE REVIEW

File: UWPPackageInstaller/App.xaml.cs [hunk 1]
```diff
@@ -132,6 +125,45 @@ protected override void OnFileActivated(FileActivatedEventArgs e)
...
+                    // just ensure that the window is active
+                    if (rootFrame == null)
+                    {
+                        // Create a Frame to act as the navigation context and navigate to the first page
+                        rootFrame = new Frame();
+
+                        rootFrame.NavigationFailed += OnNavigationFailed;
+
+
+
```
Comment (patfinzz) [inline UWPPackageInstaller/App.xaml.cs]: Remove some whitespace her please

File: UWPPackageInstaller/MainPage.xaml.cs [hunk 2]
```diff
@@ -1,385 +1,160 @@
...
-                installButton.Visibility = Visibility.Collapsed;
-                cancelButton.Content = "Exit";
-                packageNameTextBlock.Text = "No package Selected";
+                var inputSasUrl = new Uri(uriParam.OriginalString.Remove(0, "echoinstaller://".Length));
 
-            }
-        }
+                if (!EchoUrlValidator.IsUrlValidForEcho(inputSasUrl))
+                {
+                    PermissionTextBlock.Text = "Input url is invalid. Cannot install this app.";
```
Comment (patfinzz) [inline UWPPackageInstaller/MainPage.xaml.cs]: Maybe add what a correct url looks like?

File: UWPPackageInstaller/MainPage.xaml.cs [hunk 3]
```diff
@@ -1,385 +1,160 @@
...
+                PermissionTextBlock.Text = $"Install {filename}?";
+                InstallProgressBar.Visibility = Visibility.Collapsed;
+                InstallValueTextBlock.Visibility = Visibility.Collapsed;
+                InstallButton.Visibility = Visibility.Visible;
+                CancelButton.Content = "Exit";
+                PackageNameTextBlock.Text = filename;
+            }
+            else
+            {
+                // TODO: Actually create a link here, or open it directly
```
Comment (patfinzz) [inline UWPPackageInstaller/MainPage.xaml.cs]: Add link

File: UWPPackageInstaller/MainPage.xaml.cs [hunk 4]
```diff
@@ -1,385 +1,160 @@
...
-                try
-                {
-
-                    var result = await pkgManager.AddPackageAsync(new Uri(packageInContext.Path), null, DeploymentOptions.ForceTargetApplicationShutdown).AsTask(progressCallback);
-                    checkIfPackageRegistered(result, resultText);
-                }
+                var addPackageOperation = _packageManager.AddPackageAsync(fileToDownload, null,
+                    DeploymentOptions.ForceApplicationShutdown | DeploymentOptions.ForceUpdateFromAnyVersion);
+                // There is no progress callback while downloading (that I have found)
+                PermissionTextBlock.Text = "Downloading files... Download progress is unknown.";
```
Comment (patfinzz) [inline UWPPackageInstaller/MainPage.xaml.cs]: This message feels like an error message. Maybe add something like "... Install will soon start automatically" or something

**Comment by Strepto** [comment]
I think this is ready now

**Review by patfinzz — APPROVED** [review]

**Comment by Strepto** [comment]
Merged to cleanup, as this code has been used for a while

**Pull request merged**